src/amo/components/SearchFilters/index.js

```diff
@@ -228,6 +233,7 @@ export class SearchFiltersBase extends React.Component {
           </label>
           <Select
             className="SearchFilters-OperatingSystem SearchFilters-select"
+            defaultValue={convertOSToFilterValue(userAgentInfo.os.name)}
```

Comment:
So I thought that would work but I cannot make it work locally. I think that's because `defaultValue` is only used when there is no `value` at all. It is tough because what we want really is to use the `userAgentInfo` iif there is no `operatingSystem` filter but also *only* once.

STR:

- Go to http://localhost:3000/en-US/firefox/
- Use the search bar to search something
- Observe the OS filter: it should be set according to the user agent (i.e. your OS)
- Change the OS to "All"

Problem: the search results are not updated. I tried to move the `|| NO_FILTER` like this:

```diff
diff --git a/src/amo/components/SearchFilters/index.js b/src/amo/components/SearchFilters/index.js
index 2b3d7d561..0f82bcddc 100644
--- a/src/amo/components/SearchFilters/index.js
+++ b/src/amo/components/SearchFilters/index.js
@@ -233,11 +233,11 @@ export class SearchFiltersBase extends React.Component {
           </label>
           <Select
             className="SearchFilters-OperatingSystem SearchFilters-select"
-            defaultValue={convertOSToFilterValue(userAgentInfo.os.name)}
+            defaultValue={convertOSToFilterValue(userAgentInfo.os.name) || NO_FILTER}
             id="SearchFilters-OperatingSystem"
             name="operatingSystem"
             onChange={this.onSelectElementChange}
-            value={filters.operatingSystem || NO_FILTER}
+            value={filters.operatingSystem}
           >
             {this.operatingSystemOptions().map((option) => {
               return <option key={option.value} {...option} />;
```

This works a bit better but the correct `option` is not updated 😞

Reply:
Hey @willdurand , I'm a bit stuck. The cases where value of `filters.operatingSystem == ""` are working fine. But what should I do when `typeof filters.operatingSystem == "undefined"`? I was trying to change the state such that if it is undefined we can change to default value. But I'm having some troubles. Can you please help me?

Reply:
I am not sure I can because I don't fully understand your problem. I think when we have an falsey `filters.operatingSystem`, then the `defaultValue` is used but I don't think we can make it work for all cases.

Reply:
@willdurand  got some ideas from your last comment. Now it will work for all cases hopefully.

Reply:
That's interesting, thanks.